File: EpilogueIOS/Modules/GhostwriterClient/Sources/GhostwriterClient/Models/SyncModels.swift
//
//  SyncModels.swift
//  Epilogue
//
//  Combined sync response models for the GET /sync endpoint.
//

import Foundation

/// Combined sync response containing all data the client needs in one call.
public struct SyncResponse: Codable, Sendable {
    public let config: ClientConfigResponse
    public let feeds: FeedChangesResponse
    public let digests: SyncDigestsSection
    public let schedules: [ScheduleResponse]
}

/// Digests section of the sync response.
public struct SyncDigestsSection: Codable, Sendable {
    public let newDigests: [SyncDigest]

    enum CodingKeys: String, CodingKey {
        case newDigests = "new_digests"
    }
}

/// Digest with embedded articles from the sync endpoint.
public struct SyncDigest: Codable, Sendable, Identifiable {
    public let id: String
    public let filename: String
    public let period: String
    public let status: String
    public let stage: String?
    public let articleCount: Int
    public let errorMessage: String?
    public let createdAt: String
    public let completedAt: String?
    public let articles: [DigestArticleResponse]

    enum CodingKeys: String, CodingKey {
        case id, filename, period, status, stage, articles
        case articleCount = "article_count"
        case errorMessage = "error_message"
        case createdAt = "created_at"
        case completedAt = "completed_at"
    }
}

File: app/src/main/java/com/example/epilogue/data/remote/ghostwriter/SyncModels.kt
package com.example.epilogue.data.remote.ghostwriter

import com.google.gson.annotations.SerializedName

/**
 * Combined sync response from GET /sync endpoint.
 * Contains all data the client needs in a single response.
 */
data class SyncResponse(
    @SerializedName("config") val config: ClientConfigResponse,
    @SerializedName("feeds") val feeds: FeedChangesResponse,
    @SerializedName("digests") val digests: SyncDigestsSection,
    @SerializedName("schedules") val schedules: List<ScheduleResponse>
)

/**
 * Digests section of the sync response.
 */
data class SyncDigestsSection(
    @SerializedName("new_digests") val newDigests: List<SyncDigest>
)

/**
 * Digest with embedded articles from the sync endpoint.
 */
data class SyncDigest(
    @SerializedName("id") val id: String,
    @SerializedName("filename") val filename: String,
    @SerializedName("available_formats") val availableFormats: List<String>? = null,
    @SerializedName("period") val period: String,
    @SerializedName("status") val status: String,
    @SerializedName("stage") val stage: String?,
    @SerializedName("article_count") val articleCount: Int,
    @SerializedName("error_message") val errorMessage: String?,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("completed_at") val completedAt: String?,
    @SerializedName("articles") val articles: List<DigestArticleResponse>
)
